plugins/telemetry-otel/src/main/java/org/opensearch/telemetry/tracing/OTelSpan.java

```diff
@@ -32,9 +38,33 @@ public OTelSpan(String spanName, Span span, org.opensearch.telemetry.tracing.Spa
 
     @Override
     public void endSpan() {
+        if (isSpanOutlier()) {
```

Comment:
How that is suppose to work in case of async calls when parent is finished before the child? (example below)

```
        final Span span = tracer.startSpan(SpanBuilder.from(action, connection));
        try (SpanScope spanScope = tracer.withSpanInScope(span)) {
            TransportResponseHandler<T> traceableTransportResponseHandler = TraceableTransportResponseHandler.create(handler, span, tracer);
            sendRequestAsync(connection, action, request, options, traceableTransportResponseHandler);
        }
```

Reply:
So the parent span object is still maintained in the scope of child even if the parent has ended. What we plan to do here is mark the current child's parent and the chain above as one of the parent within the chain would still be alive. As suggested [here](https://github.com/opensearch-project/OpenSearch/issues/12315#issuecomment-2017803049) as well if the parent is still in the recording phase we will go ahead and add the attribute to the parent and mark the parent to be sampled as well. If the parent is not recording then the information about this parent will be stored in event of its parent so that we don't loose on the info of not-recording spans.

Adding the closed parent event to its parent span is still pending and will add in the next revision itself.

Reply:
I don't understand the terminology here (what does it mean for span chain to be alive), sorry about that , let me try to illustrate that better:

```
 try (ScopedSpan startScopedSpan = tracer.startScopedSpan(...)) {
           threadPool.executor(...).execute( ... )
  }
```

- since we sample 100%, all spans are sampled
- the parent span is created (`tracer.startScopedSpan`) , onSpanStart is called
- the `threadPool` captures context (current span as parent) and scheduled the work
- the the parent span is ended, onSpanEnd is called

After `onSpanEnd`, the parent span is immutable, no matter if we still have reference to in from Java code or not - any modification to it have no effect but produce warning. 

My question is: since we not be exporting the parent span (due to absence of the expected attributes), what would happen in this case?

Reply:
So here is how this would work:
- If the span is an outlier span we will add the sampled attribute to the span.
- Once the span is marked to be sampled we will mark its parent to be sampled as well. This could now have two possibilities
   - One, the parent is still recording. If that is the case we will go ahead and add the attribute to the parent as well and the parent would be sampled as well. 
   - Second, the parent is not recording/has ended in that case we mark its grand-parent if that is alive to be sampled and record this closed span information in the alive grand-parent itself. This way we will not get the span which was closed howe-ever the information about the span would still persist in its parent which will be sampled

Reply:
This is what I have suspected - we are messing up the trace in unpredictable ways (another red flag), there should be only one  possibility: the trace is recorded as it was sampled or not recorded.

Reply:
Yes you are correct we will miss the span. However to maintain the lineage we will still maintain the parent information of closed span in its parent itself.

Reply:
So here is the thing folks, I my opinion, which I shared on RFC as well:
 - if we cannot preserve span hierarchy, this is no go (the mess would be on user to deal with)
 - if we cannot follow the rules of the library we are using, this is no go

OTEL cannot solve tail sampling on library level, this is documented and acknowledged:
 -  https://opentelemetry.io/blog/2022/tail-sampling/
 - https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/tailsamplingprocessor

There are other tracing instrumentation that try to implemented sampling differently, including adaptive sampling that could tune the sampling rates based on the different factors:
 - https://github.com/openzipkin-attic/zipkin-zookeeper
 - https://medium.com/jaegertracing/adaptive-sampling-in-jaeger-50f336f4334

Please consider those as a robust and reliable alternatives, that do not exhibit flaws of this solution.

Reply:
I see where you're coming from regarding the potential gaps in the hierarchy, which users may need to address based on their choice of collector and visualization solutions. It's been recognized as a requirement to address such gaps in the Otel specification, and the community appears to be quite receptive to it. If you take a look at the Otel issues provided, you'll notice that even with these gaps, there are still ways to infer missing parent information through parent span events, aiding in understanding the code path and other relevant details. 

https://github.com/open-telemetry/opentelemetry-specification/issues/3205
https://github.com/open-telemetry/opentelemetry-specification/issues/3867

In general, there are numerous benefits to this approach, particularly for customers dealing with larger clusters processing tens of thousands of requests per second and managing multiple clusters. It offers insights into problematic areas of code, and the availability of missing span information as part of the event in the parent span can be very helpful in connecting the dots. 

Moreover, this won't be the default behavior and can be enabled as needed (though it's crucial to clearly document the shortcomings and benefits of such samplers). There have been requests from Otel users for deferred sampling support and reverse propagation of context, and it's hoped that these features will be available soon, at which point a switch can be made.

Reply:
One last thing, I think we can solve this issue by adding a link between the span and it's grand parent to persist the relationship. Since this is a `follows_from` span.